### ISSUE TITLE
[Minor] Early stopping for patience == 0

### ISSUE THREAD

**Issue by janLuke** [allenai/allennlp#1267]
Should 0 be a valid value for ``patience``? It looks like a reasonable value for expressing "no patience".

* If yes, I think it should be equivalent to ``patience == 1``: the training should be stopped at the end of the first epoch without improvement. 
* if no, maybe it's the case to add a check ``patience > 0`` in the constructor.

Currently, for ``patience == 0``, the stop condition is never met because in ``Trainer._should_stop_early``:
```
if self._validation_metric_decreases:
    return min(metric_history[-self._patience:]) > min(metric_history)
else:
    return max(metric_history[-self._patience:]) < max(metric_history)
```
 ``metric_history[-0:] == metric_history`` so the inequality between their min/max is always false.

**Comment by schmmd** [comment]
The current behavior (`patience = 0` means the training will never stop early due to patience) makes sense to me.  Otherwise, how would you disable patience if you wanted to?

**Comment by janLuke** [comment]
@schmmd Negative patience, for example.

**Comment by nelson-liu** [comment]
to play devil's advocate, I think `patience = 0` could just as easily be construed as "no patience" / "impatient", and that you'd stop as soon as you see an epoch where the metric doesn't improve.

fwiw, i think `patience=0` has the above behavior in Keras, but it's a bit different from an API design perspective since there you have to explicitly add an `EarlyStoppingCallback`, and thus it doesn't make sense to have a value that disables patience when you've already explicitly added this callback.

**Comment by schmmd** [comment]
I would find it confusing if `patience == 0` were the same as `patience == 1`, so maybe a reasonable proposal is that `patience` cannot be 0 and is disabled with -1?

**Comment by janLuke** [comment]
@nelson-liu I made exactly the same reasoning about Keras API. Still, I think it's simply counterintuitive that ``patience = 0`` disables  early stopping and the fact that other libraries behave differently only add to that. Furthermore, there's not a strict need to reserve zero for that function.

@schmmd Yes, I find that confusing too and your proposal sounds reasonable to me. If I had to implement early-stopping from scratch, I'd define ``patience`` as the max number of (consecutive) epochs with no improvement one can tolerate **without** stopping the training, so that the training is stopped after ``patience + 1`` unsuccessful epochs. In other words, it'd be the max number of consecutive postponements of early stopping. This way:
* ``patience == 0`` means what it suggests: no patience, stop at the first epoch with no improvement;
* ``patience == 1`` means you can tolerate 1 epoch with no improvement, but not 2 consecutive ones.

Anyway, this issue is just a minor detail. I would have never opened it if I hadn't reviewed https://github.com/allenai/allennlp/pull/1246 (which crashes for ``patience == 0``), given that I always use ``patience >= 2`` in practice. Whatever decision will be taken about it, it'll be fine as long as it is well-documented.

**Comment by matt-peters** [comment]
The most intuitive way (to me) to disable patience is to set it to `None`, which usually means "ignore this parameter and don't take any special action" in most other contexts in python.  Or set it to larger number then the total number of training epochs.

**Comment by janLuke** [comment]
Yes, setting ``patience >= num_epochs`` is another totally unambiguous way to disable early stopping. 
Weirdly enough, I had not thought to ``None``!
